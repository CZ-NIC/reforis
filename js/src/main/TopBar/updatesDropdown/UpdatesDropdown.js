/*
 * Copyright (C) 2020-2021 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect, useCallback } from "react";
import PropTypes from "prop-types";
import { Link } from "react-router-dom";

import {
    ForisURLs,
    useAPIGet,
    useAPIPost,
    SpinnerElement,
    Button,
    API_STATE,
    withEither,
    withSending,
} from "foris";
import API_URLs from "common/API";

UpdatesDropdown.propTypes = {
    newNotification: PropTypes.bool.isRequired,
};

export default function UpdatesDropdown({ newNotification }) {
    const [getApprovalsResponse, getApprovals] = useAPIGet(API_URLs.approvals);
    const update = getApprovalsResponse.data || {};
    useEffect(() => {
        getApprovals();
    }, [getApprovals]);

    const updateWithNotification = useCallback(() => {
        if (newNotification) {
            return getApprovals();
        }
    }, [newNotification, getApprovals]);

    useEffect(() => {
        updateWithNotification();
    }, [updateWithNotification]);

    if (update.approvable === false) {
        return null;
    }

    return (
        <div className="dropdown" data-testid="updates-dropdown">
            <DropdownContentWithSpinner
                apiState={getApprovalsResponse.state}
                update={update}
                onSuccess={getApprovals}
            />
        </div>
    );
}

DropdownContent.propTypes = {
    update: PropTypes.object.isRequired,
    onSuccess: PropTypes.func.isRequired,
};

function DropdownContent({ update, onSuccess }) {
    const [postApprovalResponse, postApproval] = useAPIPost(API_URLs.approvals);
    // Reload approvals when resolution is successful
    useEffect(() => {
        if (postApprovalResponse.state === API_STATE.SUCCESS) {
            onSuccess();
        }
    }, [onSuccess, postApprovalResponse]);

    function resolveUpdate(solution) {
        postApproval({ data: { hash: update.hash, solution } });
    }

    const updateFailed = postApprovalResponse.state === API_STATE.ERROR;
    return (
        <>
            <button type="button" className="nav-item btn btn-link">
                <i
                    className={`fa fa-sync-alt fa-lg ${
                        updateFailed ? "text-danger" : ""
                    }`.trim()}
                />
            </button>
            <div className="dropdown-menu dropdown-menu-right shadow-sm">
                <div className="dropdown-header">
                    <Link
                        to={{
                            pathname: ForisURLs.approveUpdates,
                        }}
                    >
                        <h5>{_("Approve Update")}</h5>
                    </Link>
                </div>
                <div className="dropdown-divider" />
                <ManageUpdateWithError
                    updateFailed={updateFailed}
                    resolveUpdate={resolveUpdate}
                />
            </div>
        </>
    );
}

const withSmallSpinner = withSending(() => (
    <button type="button" className="nav-item btn btn-link">
        <SpinnerElement small />
    </button>
));
const DropdownContentWithSpinner = withSmallSpinner(DropdownContent);

ManageUpdate.propTypes = {
    resolveUpdate: PropTypes.func.isRequired,
};

function ManageUpdate({ resolveUpdate }) {
    return (
        <>
            <span
                className="dropdown-item"
                dangerouslySetInnerHTML={{
                    __html: _(
                        `See details in <a href=${ForisURLs.packageManagement.updates}>Updates</a> page.`
                    ),
                }}
            />
            <div className="dropdown-item text-center">
                <Button
                    className="btn-primary"
                    onClick={() => resolveUpdate("grant")}
                >
                    {_("Install now")}
                </Button>
            </div>
        </>
    );
}

const withUpdateFailed = withEither(
    (props) => props.updateFailed,
    () => (
        <span className="dropdown-item text-danger">
            {_("Cannot install updates.")}
        </span>
    )
);
const ManageUpdateWithError = withUpdateFailed(ManageUpdate);
