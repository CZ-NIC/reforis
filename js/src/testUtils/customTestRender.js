/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

/* eslint import/export: "off" */

import React from 'react';
import {UIDReset} from 'react-uid';
import {StaticRouter} from 'react-router';
import {render} from '@testing-library/react'

function Wrapper({children}) {
    return <StaticRouter>
        <UIDReset>
            {children}
        </UIDReset>
    </StaticRouter>
}

const customTestRender = (ui, options) => render(ui, {wrapper: Wrapper, ...options});

// re-export everything
export * from '@testing-library/react'

// override render method
export {customTestRender as render}
