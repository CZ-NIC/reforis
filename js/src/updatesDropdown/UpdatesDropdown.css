#updates-dropdown-actions {
    display: flex;
    justify-content: space-between;
}
