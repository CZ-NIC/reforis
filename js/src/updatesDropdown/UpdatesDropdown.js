/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React, { useEffect } from "react";

import {
    ForisURLs, useAPIGet, useAPIPost, SpinnerElement, Button,
} from "foris";
import API_URLs from "common/API";

export default function UpdatesDropdown() {
    const [getState, get] = useAPIGet(API_URLs.approvals);
    const update = getState.data;
    useEffect(() => {
        get();
    }, [get]);

    const [postState, post] = useAPIPost(API_URLs.approvals);
    // Reload approvals when resolution is successful
    useEffect(() => {
        if (postState.isSuccess) {
            get();
        }
    }, [get, postState]);

    function resolveUpdate(solution) {
        post({ hash: update.hash, solution });
    }

    if (getState.isLoading || !update) {
        return (
            <div className="dropdown" id="updates-dropdown">
                <button type="button" className="nav-item btn btn-link">
                    <SpinnerElement small />
                </button>
            </div>
        );
    }

    if (!update.approvable) {
        return null;
    }

    let hasError = false;
    let dropdownContent;
    if (postState.isError) {
        hasError = true;
        dropdownContent = <span className="dropdown-item text-danger">{_("Cannot resolve update")}</span>;
    } else {
        dropdownContent = (
            <>
                <span
                    className="dropdown-item"
                    dangerouslySetInnerHTML={{ __html: _(`See details in <a href=${ForisURLs.packageManagement.updates}>Updates</a> page.`) }}
                />
                <div className="dropdown-item" id="updates-dropdown-actions">
                    <Button className="btn-warning mr-3" onClick={() => resolveUpdate("deny")}>{_("Ignore")}</Button>
                    <Button className="btn-primary" onClick={() => resolveUpdate("grant")}>{_("Install now")}</Button>
                </div>
            </>
        );
    }

    return (
        <div className="dropdown">
            <button type="button" className="nav-item btn btn-link">
                <i className={`fa fa-sync fa-lg ${hasError ? "text-danger" : ""}`.trim()} />
            </button>
            <div className="dropdown-menu">
                <div className="dropdown-header">
                    <h5>{_("Approve update")}</h5>
                </div>
                <div className="dropdown-divider" />
                {dropdownContent}
            </div>
        </div>
    );
}
