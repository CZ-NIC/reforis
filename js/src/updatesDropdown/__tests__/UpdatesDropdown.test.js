/*
 * Copyright (C) 2019 CZ.NIC z.s.p.o. (http://www.nic.cz/)
 *
 * This is free software, licensed under the GNU General Public License v3.
 * See /LICENSE for more information.
 */

import React from "react";
import { render, waitForElement, waitForElementToBeRemoved, wait, fireEvent } from "customTestRender";
import mockAxios from 'jest-mock-axios';
import { mockJSONError } from "foris";

import UpdatesDropdown from "../UpdatesDropdown";

describe("<UpdatesDropdown/>", () => {
    let container, getByText;
    const exampleHash = "303808909";

    beforeEach(() => {
        ({ container, getByText } = render(<UpdatesDropdown />));
    });

    it("Loading (spinner visible)", async () => {
        expect(container).toMatchSnapshot();
    });

    it("No updates awaiting", async () => {
        mockAxios.mockResponse({data: {approvable: false}});
        await waitForElementToBeRemoved(() => document.querySelector('#updates-dropdown'));
    });

    it("Updates awaiting - snapshot", async () => {
        mockAxios.mockResponse({data: {hash: exampleHash, approvable: true}});
        await waitForElement(() => getByText("Approve update"));
        expect(container).toMatchSnapshot();
    });

    it("Updates awaiting - install now", async () => {
        mockAxios.mockResponse({data: {hash: exampleHash, approvable: true}});
        await waitForElement(() => getByText("Approve update"));

        fireEvent.click(getByText("Install now"));
        expect(mockAxios.post).toBeCalled();
        expect(mockAxios.post).toHaveBeenCalledWith("/api/approvals", {"hash": exampleHash, "solution": "grant"}, expect.anything());

        // Reload approvals when resolution is successful
        await wait(() => expect(mockAxios.get).toBeCalledTimes(1));
        mockAxios.mockResponse({data: {}});
        await wait(() => expect(mockAxios.get).toBeCalledTimes(2));
    });

    it("Updates awaiting - ignore", async () => {
        mockAxios.mockResponse({data: {hash: exampleHash, approvable: true}});
        await waitForElement(() => getByText("Approve update"));

        fireEvent.click(getByText("Ignore"));
        expect(mockAxios.post).toBeCalled();
        expect(mockAxios.post).toHaveBeenCalledWith("/api/approvals", {"hash": exampleHash, "solution": "deny"}, expect.anything());
    });

    it("Updates resolution - display error", async () => {
        mockAxios.mockResponse({data: {hash: exampleHash, approvable: true}});
        await waitForElement(() => getByText("Approve update"));
        fireEvent.click(getByText("Install now"));
        mockJSONError();
        await waitForElement(() => getByText("Cannot resolve update"));
    });
});
