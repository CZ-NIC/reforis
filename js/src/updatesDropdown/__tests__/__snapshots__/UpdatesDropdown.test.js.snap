// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<UpdatesDropdown/> Loading (spinner visible) 1`] = `
<div>
  <div
    class="dropdown"
    data-testid="updates-dropdown"
  >
    <button
      class="nav-item btn btn-link"
      type="button"
    >
      <div
        class="spinner-border spinner-border-sm"
        role="status"
      >
        <span
          class="sr-only"
        />
      </div>
      <div
        class="spinner-text"
      />
    </button>
  </div>
</div>
`;

exports[`<UpdatesDropdown/> Updates awaiting - snapshot 1`] = `
<div>
  <div
    class="dropdown"
    data-testid="updates-dropdown"
  >
    <button
      class="nav-item btn btn-link"
      type="button"
    >
      <i
        class="fa fa-sync fa-lg"
      />
    </button>
    <div
      class="dropdown-menu"
    >
      <div
        class="dropdown-header"
      >
        <h5>
          Approve update
        </h5>
      </div>
      <div
        class="dropdown-divider"
      />
      <span
        class="dropdown-item"
      >
        See details in 
        <a
          href="/reforis/package-management/updates"
        >
          Updates
        </a>
         page.
      </span>
      <div
        class="dropdown-item"
        id="updates-dropdown-actions"
      >
        <button
          class="btn btn-warning mr-3"
          type="button"
        >
           
           
          Ignore
        </button>
        <button
          class="btn btn-primary"
          type="button"
        >
           
           
          Install now
        </button>
      </div>
    </div>
  </div>
</div>
`;
